### PR TITLE
RFC: join framework design

### PR DESCRIPTION
**Pull request by leiysky**
Signed-off-by: leiysky <leiysky@outlook.com>

## Summary

Proposal for Join.

## Changelog

- New Feature
- Documentation

## Related Issues
#319 

## Test Plan

No test.

### CODE REVIEW

**Comment by leiysky** [comment]
https://github.com/leiysky/fuse-query/blob/rfc/join-framework-design/docs/rfcs/query/2021-05-01-join-framework-design.md

Rendered document.

**Comment by BohuTANG** [comment]
@leiysky  

Although this is the initial version, it is very helpful, thank you for bringing it to datafuse.

File: docs/rfcs/query/2021-05-01-join-framework-design.md [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+- `INNER JOIN`: return all tuples satisfy the join condition
+- `LEFT OUTER JOIN`: return all tuples satisfy the join condition and the rows from left table for which no row from right table satisfies the join condition
+- `RIGHT OUTER JOIN`: return all tuples satisfy the join condition and the rows from right table for which no row from left table satisfies the join condition
+- `FULL OUTER JOIN`: return all tuples satisfy the join condition and the rows from a table for which no row from other table satisfies the join condition
+- `CROSS JOIN`: cartesian product of joined tables
+
+Besides, `IN`, `EXISTS`, `NOT IN`, `NOT EXISTS` expressions can be implemented by **semi-join** and **anti-join**(known as subquery).
+
+There are kinds of join algorithms:
```
Comment (jyizheng) [inline docs/rfcs/query/2021-05-01-join-framework-design.md:19]: miss "three" here?

Comment (leiysky) [reply docs/rfcs/query/2021-05-01-join-framework-design.md]: Fixed.

**Comment by tomjiang1987** [comment]
the initial rfc is fine, and also need change the exist code to make it consistent with the new design.

**Comment by leiysky** [comment]
@BohuTANG I'd like to know about the standard process of developing a feature.

IMO, we can merge this pull request now and keep improving it. At the same time we can list development tasks and track them in #319.

**Review by BohuTANG — APPROVED** [review]
LGTM
Let's merge, thank you @leiysky

**Comment by BohuTANG** [comment]
> @BohuTANG I'd like to know about the standard process of developing a feature.
> 
> IMO, we can merge this pull request now and keep improving it. At the same time we can list development tasks and track them in #319.

ACK @zhang2014

**Comment by BohuTANG** [comment]
Hi @leiysky ,

There are still some pre-tasks that need to be completed, such as #468 and others, and when they are almost done, we will list some specific tasks for #319 .
Cc @zhang2014

**Pull request merged**